### ISSUE TITLE
Unexpected uninitialized warning on [AllowNull] field

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#37138]
**Version Used**: 
3.3.0-beta1-19354-02+c82dbc6503182784647003a4fce50ef56d680037

**Steps to Reproduce**:
```C#
#nullable enable

using System.Diagnostics.CodeAnalysis;

class Program
{
    static void Main() { }
}

class C<T>
{
    [AllowNull] public T Value;

    public C() { }
}
```

**Expected Behavior**:
No warning.

**Actual Behavior**:
Produces a warning "Non-nullable field 'Value' is uninitialized".  But adding `= default` on `Value` suppresses that warning.  These are semantically equivalent, so it's odd that there's a warning without the `= default`.

**Comment by jcouv** [comment]
Relates to https://github.com/dotnet/roslyn/issues/36039: Attributes only affect callers at the moment, not implementation bodies or OHI.

**Comment by sharwell** [comment]
It doesn't make sense for `AllowNull` to trigger this. `AllowNull` only affects assignments to the variable, where the warning in question related to the state for reads. The `MaybeNull` attribute is the one that should suppress the warning.

**Comment by stephentoub** [comment]
> It doesn't make sense for AllowNull to affect this behavior.

I think it does.  If you write:
```C#
[AllowNull] public T Value = default;
```
the compiler will elide the `= default;` from the generated IL, because the `= default` is unnecessary and duplicative with the implicit default value that's used to initialize the field.  Yet if I leave off the `= default`:
```C#
[AllowNull] public T Value;
```
I get a warning about it being uninitialized, even though they result in exactly the same IL and are equally initialized.

**Comment by sharwell** [comment]
Why would you want `AllowNull` on a field but not have `MaybeNull` on the same field? I can see cases where `AllowNull` alone would work on a property, but for a field it seems like you are trying to write `T??` using the attribute equivalents.

**Comment by stephentoub** [comment]
> Why would it make sense to have AllowNull on a field but not have MaybeNull on the same field?

Having both makes sense; didn't say otherwise.

**Comment by stephentoub** [comment]
As for why it would make sense, it's effectively serving the same purpose as `= default!`.
